File: 039_BinaryTreeRightSideView.py
"""
Topic: Tree / BFS
Binary Tree Right Side View: collect rightmost node at each level via BFS
Time Complexity: O(n)
Space Complexity: O(n)
"""

from typing import Optional, List
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def right_side_view(root: Optional[TreeNode]) -> List[int]:
    if not root: return []
    q = deque([root])
    res = []
    while q:
        sz = len(q)
        for i in range(sz):
            node = q.popleft()
            if i == sz-1: res.append(node.val)
            if node.left: q.append(node.left)
            if node.right: q.append(node.right)
    return res

if __name__ == '__main__':
    root = TreeNode(1, TreeNode(2, None, TreeNode(5)), TreeNode(3, None, TreeNode(4)))
    print(right_side_view(root))  # [1,3,4]
    print(right_side_view(None))  # []

File: 039_BinaryTreeRightSideView.cpp
// Topic: Binary Tree / BFS
// Binary Tree Right Side View: List nodes visible when looking from the right.
// Time Complexity: O(n)
// Space Complexity: O(n)

#include <bits/stdc++.h>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

vector<int> rightSideView(TreeNode* root)
{
    vector<int> res;
    if (!root)
        return res;
    queue<TreeNode*> q;
    q.push(root);
    while (!q.empty())
    {
        int sz = q.size();
        for (int i = 0; i < sz; ++i)
        {
            auto node = q.front();
            q.pop();
            if (i == sz - 1)
                res.push_back(node->val);
            if (node->left)
                q.push(node->left);
            if (node->right)
                q.push(node->right);
        }
    }
    return res;
}

int main()
{
    TreeNode* r = new TreeNode(1);
    r->right = new TreeNode(3);
    r->left = new TreeNode(2);
    r->left->right = new TreeNode(5);
    r->right->right = new TreeNode(4);
    auto v = rightSideView(r);
    for (int x : v) cout << x << ' ';
    cout << "\n";  // 1 3 4
    TreeNode* r2 = new TreeNode(1);
    cout << rightSideView(r2)[0] << "\n";
    return 0;
}
